### Import required dependencies

In [26]:
import pandas as pd
pd.options.mode.chained_assignment = None  # default='warn'
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [27]:
# Create the path and import the data
full_student_data = os.path.join("Resources", "new_full_student_data.csv")
student_df = pd.read_csv(full_student_data)

In [28]:
# Verify that the data was properly imported
print (student_df)

       student_id      student_name grade             school_name  \
0       103880842     Travis Martin   9th    Sullivan High School   
1        45069750     Michael Brown   9th       Dixon High School   
2        45024902   Gabriela Lucero   9th      Wagner High School   
3        62582498  Susan Richardson   9th       Silva High School   
4        16437227      Sherry Davis  11th      Bowers High School   
...           ...               ...   ...                     ...   
19509   109236636     Robert Sawyer  12th       Silva High School   
19510    63239258      David Herman   9th       Woods High School   
19511    95516554        Megan Gill  11th      Wagner High School   
19512    65050383        Lori Stone  11th      Bowers High School   
19513    34720657       Anna Jensen  12th  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter   

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [29]:
# Check for null values
student_df.isnull().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [30]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df.isnull().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [31]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [32]:
# Drop duplicated rows and verify removal
student_df = student_df.drop_duplicates()
student_df.duplicated().sum()

0

In [33]:
# Check data types
print(student_df.dtypes)

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object


In [34]:
# Examine the grade column to understand why it is not an int
print(student_df['grade'])

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object


In [35]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade'] = student_df['grade'].str.replace("th", "") 
print(student_df['grade'])

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object


In [36]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].astype(str).astype(int)
print(student_df.dtypes)

student_id         int64
student_name      object
grade              int32
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object


In [37]:
#Use the head (and/or the tail) function to preview the DataFrame.
print(student_df.head().append(student_df.tail()))

       student_id      student_name  grade             school_name  \
0       103880842     Travis Martin      9    Sullivan High School   
1        45069750     Michael Brown      9       Dixon High School   
2        45024902   Gabriela Lucero      9      Wagner High School   
3        62582498  Susan Richardson      9       Silva High School   
5        74579444   Cynthia Johnson      9  Montgomery High School   
19508    83985333   Deborah Sanders     10       Silva High School   
19509   109236636     Robert Sawyer     12       Silva High School   
19511    95516554        Megan Gill     11      Wagner High School   
19512    65050383        Lori Stone     11      Bowers High School   
19513    34720657       Anna Jensen     12  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Publi

C:\Users\aryaj\AppData\Local\Temp\ipykernel_5808\884200504.py:2: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  print(student_df.head().append(student_df.tail()))


## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [38]:
# Display summary statistics for the DataFrame
print(student_df.describe())

         student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000


In [39]:
# Display the mean math score using the mean function
print(student_df['math_score'].mean())

64.67573326141189


In [40]:
# Store the minimum reading score as min_reading_score
min_reading_score = (student_df['reading_score'].min())
print(min_reading_score)

10.5


## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [41]:
# Use loc to display the grade column
print(student_df.loc[:, 'grade'])

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int32


In [42]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
print(student_df.iloc[:, [2,3,4]].head(3))

   grade           school_name  reading_score
0      9  Sullivan High School           59.0
1      9     Dixon High School           94.7
2      9    Wagner High School           89.0


In [43]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
student_df.loc[student_df['grade'] == 9].describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [44]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_row = student_df.loc[student_df['reading_score'] == min_reading_score]
print(min_reading_row)

      student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334  


In [45]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
#print(student_df.loc[(student_df['grade'] == 10) & (student_df['school_name'] == 'Dixon High School')])
print(student_df.loc[(student_df['grade'] == 10) & (student_df['school_name'] == 'Dixon High School'), 'reading_score'])

45       71.1
60       59.5
69       88.6
94       81.5
100      95.3
         ... 
19283    52.9
19306    58.0
19344    38.0
19368    84.4
19445    43.9
Name: reading_score, Length: 569, dtype: float64


In [46]:
# Find the mean reading score for all students in grades 11 and 12 combined.
print(student_df.loc[(student_df['grade'] >= 11), 'reading_score'].mean())

74.90038089192188


## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [47]:
# Use groupby and mean to find the average reading and math scores for each school type.
print(student_df.groupby(['school_type'])['reading_score'].mean())
print(student_df.groupby(['school_type'])['math_score'].mean())

school_type
Charter    72.450603
Public     72.281219
Name: reading_score, dtype: float64
school_type
Charter    66.761883
Public     62.951576
Name: math_score, dtype: float64


In [48]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
#print(student_df.groupby(['school_name'])['school_name'].count())
print(student_df.groupby(['school_name'])['school_name'].count().reset_index(name='count').sort_values(['count'], ascending = False))


               school_name  count
7   Montgomery High School   2038
6        Green High School   1961
4        Dixon High School   1583
13      Wagner High School   1541
10       Silva High School   1109
14       Woods High School   1052
11    Sullivan High School    971
12      Turner High School    846
0       Bowers High School    803
5       Fisher High School    798
9      Richard High School    551
2       Campos High School    541
8     Odonnell High School    459
1     Campbell High School    407
3        Chang High School    171


In [49]:
#Using the groupby and mean functions, find the average budget per grade for each school type.

print(student_df.groupby(['school_type', 'grade'])['reading_score'].mean())

school_type  grade
Charter      9        68.676790
             10       69.628822
             11       80.596379
             12       70.475421
Public       9        69.687981
             10       73.160461
             11       73.486448
             12       73.342380
Name: reading_score, dtype: float64


# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

While performing this analysis, I was able to identify that charter schools, on average, were found to have higher reading and math scores than public schools, with a mean score of 72.450603 for charter shcools reading score opposed to Public schools mean score of 72.281219. As for math scores, Charter schools had an average of four points higher than public schools. This dataset helps us determine that students attenting charter schools are on average more likely to score higher in reading and math. 